drivers/adc/adc_w91.c

```diff
@@ -0,0 +1,343 @@
+
+	return err;
+}
+
+/* APIs implementation: ADC read */
+static size_t pack_adc_w91_ipc_read(uint8_t inst, void *unpack_data, uint8_t *pack_data)
+{
+	struct adc_read_req *p_adc_read = unpack_data;
+	size_t pack_data_len =
+		sizeof(uint32_t) + sizeof(p_adc_read->ch) + sizeof(p_adc_read->len);
+
+	if (pack_data != NULL) {
+		uint32_t id = IPC_DISPATCHER_MK_ID(IPC_DISPATCHER_ADC_READ, inst);
+
+		IPC_DISPATCHER_PACK_FIELD(pack_data, id);
+		IPC_DISPATCHER_PACK_FIELD(pack_data, p_adc_read->ch);
+		IPC_DISPATCHER_PACK_FIELD(pack_data, p_adc_read->len);
+	}
+	return pack_data_len;
+}
+
+IPC_DISPATCHER_UNPACK_FUNC_ONLY_WITH_ERROR_PARAM(adc_w91_ipc_read);
+
+static int adc_w91_ipc_read(struct adc_w91_data *data, uint16_t ch,
+				   uint32_t len)
+{
+	struct adc_read_req adc_req = {.ch = ch, .len = len};
+	int err;
+
+	struct ipc_based_driver *ipc_data = &data->ipc;
+
+	IPC_DISPATCHER_HOST_SEND_DATA(ipc_data, data->instance_id, adc_w91_ipc_read, &adc_req, &err,
+				      CONFIG_TELINK_W91_IPC_DISPATCHER_TIMEOUT_MS);
+	if (err != 0) {
+		LOG_ERR("ADC read failed, ret(%d)", err);
+	}
+
+	return err;
+}
+
+/* Validate ADC data buffer size */
+static int adc_w91_validate_buffer_size(const struct adc_sequence *sequence)
+{
+	size_t needed = sizeof(int16_t);
+
+	if (sequence->options) {
+		needed *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed) {
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+/* Validate ADC read API input parameters */
+static int adc_w91_validate_sequence(const struct adc_sequence *sequence)
+{
+	int status;
+
+	if (sequence->channels != BIT(0)) {
+		LOG_ERR("Only channel 0 is supported.");
+		return -ENOTSUP;
+	}
+
+	if (sequence->oversampling) {
+		LOG_ERR("Oversampling is not supported.");
+		return -ENOTSUP;
+	}
+
+	status = adc_w91_validate_buffer_size(sequence);
+	if (status) {
+		LOG_ERR("Buffer size too small.");
+		return status;
+	}
+
+	return 0;
+}
+
+/* ADC Context API implementation: start sampling */
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_w91_data *data =
+	CONTAINER_OF(ctx, struct adc_w91_data, ctx);
+
+	data->repeat_buffer = data->buffer;
+	adc_w91_ipc_read(data, data->channel, 1);
+}
+
+/* ADC Context API implementation: buffer pointer */
+static void adc_context_update_buffer_pointer(struct adc_context *ctx, bool repeat_sampling)
+{
+		struct adc_w91_data *data =
+		CONTAINER_OF(ctx, struct adc_w91_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+/* Start ADC measurements */
+static int adc_w91_adc_start_read(const struct device *dev, const struct adc_sequence *sequence)
+{
+	int status;
+	struct adc_w91_data *data = dev->data;
+	const struct adc_w91_cfg *cfg = dev->config;
+
+	/* Validate input parameters */
+	status = adc_w91_validate_sequence(sequence);
+	if (status != 0) {
+		return status;
+	}
+
+	/* Check resolution */
+	if (sequence->resolution != 12) {
```

Comment:
Previous you add checker as _adc_w91_validate_sequence_ but now you are doing it directly. Please do all stuffs in same way.

Reply:
fixed